### ISSUE TITLE
Implement release mode for Linux

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#38478]
Currently a release build (from the `flutter` tool perspective) is using a debug library build, in JIT mode; the only part that's different is anything at the Runner level (e.g., build flags for the native Runner code).

We should be building and uploading a release build of the engine library, and using that with AOT for release builds, as on mobile.

**Comment by jonahwilliams** [comment]
I'm think we'll be able to use the same `.so` output for AOT we use for Android

**Comment by jld3103** [comment]
Is there a timeline for when this gets implemented? That would be great.

**Comment by stuartmorgan** [comment]
There is no timeline. Given that the Linux embedding still needs to be completely rewritten on top of something other than GLFW, release build support is not currently a priority.

**Comment by jld3103** [comment]
I'm coming from go-flutter. There we think go-flutter and also flutter-rs would profit from release mode.

**Comment by stuartmorgan** [comment]
If you are using go-flutter then this bug is irrelevant to you. It covers things like building and posting release builds of the Linux embedding library, wiring up AOT in that embedding, adding the relevant paths in the `flutter` tool, and updating the runner makefile; none of those things would apply to go-flutter or flutter-rs.

If you believe there is something preventing you from implementing AOT support in your embedding you should file a bug about that.

**Comment by jld3103** [comment]
The problem is that there is no AOT compiled version of the engine available to download. Or I missed something in the wiki.

**Comment by pchampio** [comment]
@jld3103 This issue isn't related to the build-bot or wiki doc.
Please file another bug for that regard.

**Comment by jld3103** [comment]
Oh I understand

**Comment by stuartmorgan** [comment]
> there is no AOT compiled version of the engine available to download

Implementing release mode support for the Linux desktop embedding will not change that. Again, please file a new issue.

**Comment by ping996** [comment]
does the release of AOT for linux embedding has released?

**Comment by stuartmorgan** [comment]
No; this issue will be closed when there is a release mode implemented for Linux.

**Comment by ping996** [comment]
How to use the command of `flutter build bundle` to build an AOT type one?

**Comment by stuartmorgan** [comment]
Blocked by #50778

**Comment by MarcusTomlinson** [comment]
\o

**Comment by stuartmorgan** [comment]
@MarcusTomlinson FYI the tool side of this would require either uploading release builds of the GLFW embedding, or switch to GTK first (which I've recently made upload release builds), so for now it'll require running with `--local-engine`

**Comment by h312903294** [comment]
Do we have already support linux release? I have tried to compile a linux release verison, but it coredump when it started

**Comment by h312903294** [comment]
I hava compiled a flutter app on linux, I got a directory named flutter_assets, it contains vm_snapshot_data, isolate_snapshot_data, kernel_blob.bin and so on. if there is a way that I can wrapp all of these files to a shared library named app_elf_snapshot.so. I havae used a tool named gen_snapshot, but it triggled a release_assert

**Comment by stuartmorgan** [comment]
@h312903294 Release mode isn't supported yet; that's what this issue is tracking. Unless you're building your engine locally, and have recreated (or found and patched in) the tool work to do a release assemble, you don't really have any of the pieces you need for release mode.

**Comment by h312903294** [comment]
@stuartmorgan thank for your replying. May I ask when will release mode  be ready to use? :smile:

**Comment by MarcusTomlinson** [comment]
@h312903294, Soon™ ;) Will PR for this today.

**Comment by h312903294** [comment]
@MarcusTomlinson WoW, wait for this PR

**Comment by MarcusTomlinson** [comment]
@NicholasHwang, there are a number of moving parts here though, so don't expect too swift a turnaround. It is being worked on, that much you can be sure of :)

**Comment by NicholasHwang** [comment]
@MarcusTomlinson thank for your replying and all you guys working

**Comment by NicholasHwang** [comment]
@MarcusTomlinson  I pull your commit in #57135 , and I have tried flutter build bundle, but there is no  libapp.so generated, and I have tried flutter assemble command, but I don't know how to use this command. so  if you can give some advices to me about build operation on linux release mode? thanks a lot.

**Comment by MarcusTomlinson** [comment]
@NicholasHwang, you won’t be able to use that branch yet. As I mentioned before, there are a number of moving parts here, some of which are entirely missing still. I suggest waiting until this issue is actually closed before considering it ready to use.

**Comment by liudonghua123** [comment]
I updated to the latest master branch, and build failed with the following errors.

```shell
Run flutter build linux
Downloading linux-x64/linux-x64-flutter-glfw tools...               1.4s
Downloading linux-x64/flutter-cpp-client-wrapper-glfw tools...         0.2s
Downloading linux-x64/linux-x64-flutter-gtk tools...                1.0s
🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 
Warning: Only debug is currently implemented for Linux. This is effectively a debug build.
See https://github.com/flutter/flutter/issues/38478 for details and updates.
🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 🚧 

Building Linux application...                                   
CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.

CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
Unable to generate build files
##[error]Process completed with exit code 1.
```

**Comment by MarcusTomlinson** [comment]
> I updated to the latest master branch, and build failed with the following errors.

This issue is still open, thus not ready for use yet. I suggest waiting until this issue is closed before trying profile or release builds on your project.

**Comment by hpoul** [comment]
should `flutter build linux --release` already work, or is there still something missing? I've tried on the latest master channel and got some errors which make it look like the flutter engine binaries (and `gen_snapshot`) is missing. (`flutter run` works)

<details><summary>1.19.0-2.0.pre.213, on Linux</summary>

```
[✓] Flutter (Channel master, 1.19.0-2.0.pre.213, on Linux, locale en_US.UTF-8)
    • Flutter version 1.19.0-2.0.pre.213 at /code/flutter
    • Framework revision eb1a6efe16 (7 hours ago), 2020-05-30 21:58:02 -0400
    • Engine revision 923687f0e7
    • Dart version 2.9.0 (build 2.9.0-13.0.dev 6489a0c68d)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, set ANDROID_SDK_ROOT to that location.
      You may also want to add it to your PATH environment variable.

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] Connected device (1 available)
    • Linux • Linux • linux-x64 • Linux
```
</details>

<details><summary>Flutter crash report: Cannot find executable</summary>

```

Please report a bug at https://github.com/flutter/flutter/issues.

## command

flutter --verbose assemble --output=build -dTargetPlatform=linux-x64 -dTrackWidgetCreation=true -dBuildMode=release -dTargetFile=lib/env/production.dart -dTreeSh
akeIcons="true" -dDartObfuscation=false release_bundle_linux_assets

## exception

ArgumentError: Invalid argument(s): Cannot find executable for /code/flutter/bin/cache/artifacts/engine/linux-x64-release/gen_snapshot.

#0      _getExecutable (package:process/src/interface/local_process_manager.dart:127:5)
#1      LocalProcessManager.start (package:process/src/interface/local_process_manager.dart:43:30)
#2      _DefaultProcessUtils.start (package:flutter_tools/src/base/process.dart:462:28)
#3      _DefaultProcessUtils.stream (package:flutter_tools/src/base/process.dart:480:35)
#4      GenSnapshot.run (package:flutter_tools/src/base/build.dart:74:26)
#5      AOTSnapshotter.build (package:flutter_tools/src/base/build.dart:198:56)
#6      AotElfBase.build (package:flutter_tools/src/build_system/targets/dart.dart:296:52)
#7      _BuildInstance._invokeInternal (package:flutter_tools/src/build_system/build_system.dart:795:27)
#8      _rootRunUnary (dart:async/zone.dart:1198:47)
#9      _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#10     _FutureListener.handleValue (dart:async/future_impl.dart:143:18)
#11     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:696:45)
#12     Future._propagateToListeners (dart:async/future_impl.dart:725:32)
#13     Future._completeWithValue (dart:async/future_impl.dart:529:5)
#14     Future._asyncCompleteWithValue.<anonymous closure> (dart:async/future_impl.dart:567:7)
#15     _rootRun (dart:async/zone.dart:1190:13)
#16     _CustomZone.run (dart:async/zone.dart:1093:19)
#17     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
#18     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
#19     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#20     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#21     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#22     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:169:5)
```
</details>


The build output seems to indicate it is looking for `flutter/bin/cache/artifacts/engine/linux-x64-release` which does not exist:
```
[        ] [  +35 ms] executing: /code/flutter/bin/cache/artifacts/engine/linux-x64-release/gen_snapshot --deterministic
--snapshot_kind=app-aot-elf --elf=/code/authpass/authpass/.dart_tool/flutter_build/e31eed6f85cb4dcc330be305ccd51c6c/app.so
--strip --no-causal-async-stacks --lazy-async-stacks
/code/authpass/authpass/.dart_tool/flutter_build/e31eed6f85cb4dcc330be305ccd51c6c/app.dill
[        ] [  +18 ms] Persisting file store
[        ] [  +44 ms] Done persisting file store
[        ] [  +31 ms] "flutter assemble" took 57,145ms.
[        ] Oops; flutter has exited unexpectedly: "Invalid argument(s): Cannot find executable for
/code/flutter/bin/cache/artifacts/engine/linux-x64-release/gen_snapshot.".
[        ] [+4481 ms] executing: [/code/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%ar
[        ] [  +82 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%ar
[        ] [        ] 7 hours ago
[        ] [  +72 ms] executing: /code/flutter/bin/cache/artifacts/engine/android-arm-profile/linux-x64/gen_snapshot
```

this directory only contains:

```
➜  code ls -l flutter/bin/cache/artifacts/engine
total 0
drwxr-xr-x   3 herbert  staff   96 May 31 10:03 android-arm
drwxr-xr-x   4 herbert  staff  128 May 31 10:03 android-arm-profile
drwxr-xr-x   4 herbert  staff  128 May 31 10:03 android-arm-release
drwxr-xr-x   3 herbert  staff   96 May 31 10:03 android-arm64
drwxr-xr-x   4 herbert  staff  128 May 31 10:03 android-arm64-profile
drwxr-xr-x   4 herbert  staff  128 May 31 10:03 android-arm64-release
drwxr-xr-x   4 herbert  staff  128 May 31 10:02 android-x64
drwxr-xr-x   4 herbert  staff  128 May 31 10:03 android-x64-profile
drwxr-xr-x   4 herbert  staff  128 May 31 10:03 android-x64-release
drwxr-xr-x   4 herbert  staff  128 May 31 10:02 android-x86
drwxr-xr-x   3 herbert  staff   96 May 31 10:03 android-x86-jit-release
drwxr-xr-x   4 herbert  staff  128 May 30 13:44 common
drwxr-xr-x  17 herbert  staff  544 May 31 10:18 linux-x64
```

<details><summary>i have tried to run `flutter precache -f -v --linux --no-android` but this didn't change anything.</summary>
<pre>
[  +18 ms] executing: [/code/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +93 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] eb1a6efe1637b6213d94ad6e2d159a3d50e56699
[        ] executing: [/code/flutter/] git tag --contains HEAD
[ +306 ms] Exit code 0 from: git tag --contains HEAD
[   +4 ms] executing: [/code/flutter/] git describe --match *.*.*-*.*.pre --first-parent --long --tags
[ +260 ms] Exit code 0 from: git describe --match *.*.*-*.*.pre --first-parent --long --tags
[        ] 1.19.0-1.0.pre-213-geb1a6efe1
[  +12 ms] executing: [/code/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[  +67 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/master
[        ] executing: [/code/flutter/] git ls-remote --get-url origin
[  +27 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[ +155 ms] Unable to locate an Android SDK.
[  +26 ms] executing: [/code/flutter/] git rev-parse --abbrev-ref HEAD
[  +49 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] master
[ +208 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[  +10 ms] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[  +14 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[  +13 ms] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[ +100 ms] "flutter precache" took 251ms.
root@817c37e9d3f3:/code/authpass/authpass# 
</pre></summary>

Am I missing something (obvious), or should I file a new issue?

**Comment by stuartmorgan** [comment]
> Am I missing something (obvious), or should I file a new issue?

Please file a new issue.

**Comment by cachapa** [comment]
Same problem here (Ubuntu 20.04)
I'm guessing this will be tracked in #58365

**Comment by MarcusTomlinson** [comment]
> Please file a new issue.

Is this not https://github.com/flutter/flutter/issues/57248?

**Comment by stuartmorgan** [comment]
> I'm guessing this will be tracked in #58365

Yes, we'll use that. I'll have a fix shortly.

> Is this not #57248?

That is fixed, and I just never marked it as such. This is a tool bug where it's not downloading those artifacts, which I missed in review.

**Comment by NicholasHwang** [comment]
how long will this change release in a flutter version

**Comment by stuartmorgan** [comment]
Flutter's `master` channel is just the master branch of the repository, so this change was available in Flutter the moment it landed.

**Comment by limbouser** [comment]
Will I be able to generate a AOT/Release mode bundle for linux-arm? For eg: a Raspberry Pi, I have an embedder that runs the application in debug mode, it would be nice if I could switch it to Release mode.

**Comment by stuartmorgan** [comment]
This issue was specific to the official Flutter Linux embedding. If you're using a custom embedding, it's up to you to build all the necessary pieces (release build of the engine, correct gen_snapshot, AOT bundle, embedder code to provide it using the embedder API, etc.)
